Added a Model::clean function to remove unneeded items

kerimoyle opened the pull request:
Addresses #775

kerimoyle committed:
test to show intended behaviour

kerimoyle committed:
Better behaviour, function operates on itself

kerimoyle committed:
WIP computer crashing ...

kerimoyle committed:
tests added, all good

kerimoyle committed:
docstrings

kerimoyle committed:
Trying to fix windows/macos build errors

kerimoyle committed:
added include <iterator> to make windows happy (I hope)

kerimoyle committed:
Removed set_difference as behaviour is different btw OS (!?)

kerimoyle committed:
Python test for coverage

kerimoyle commented:
Ready for review please @agarny @hsorby @nickerso

kerimoyle committed:
Merge branch 'develop' into i775-model-clean-function

agarny commented on src/model.cpp:
```diff
@@ -539,4 +539,70 @@ bool Model::fixVariableInterfaces()
+        }
+    }
+
+    return (component->variableCount() + component->resetCount() + component->componentCount()) == 0
```
This should be `(component->variableCount() + component->resetCount() + component->componentCount() == 0)`.

kerimoyle replied:
f15e055d

kerimoyle committed:
Merge branch 'develop' into i775-model-clean-function

agarny committed:
Model and importer: don't cast a `size_t` to an `int`.

agarny committed:
Merge branch 'i775-model-clean-function' of https://github.com/kerimoyle/libcellml into pr/778

agarny commented on src/model.cpp:
```diff
@@ -547,4 +547,72 @@ bool Model::fixVariableInterfaces()
+        if (u != nullptr) {
+            auto it = std::find_if(unused.begin(), unused.end(), [=](const UnitsPtr &t) -> bool {
+                // KRM Not sure whether including names is a good idea or not.  They should be linked by
+                //     pointers only anyway, but the findUnits function is still testing names and equivalence ...
```
Remove.

kerimoyle replied:
Yeah ... not until I know what's going on ... For the final thing, sure.

kerimoyle replied:
Moot now as have removed this function 8837c78c

agarny commented on src/model.cpp:
```diff
@@ -547,4 +547,72 @@ bool Model::fixVariableInterfaces()
+            auto it = std::find_if(unused.begin(), unused.end(), [=](const UnitsPtr &t) -> bool {
+                // KRM Not sure whether including names is a good idea or not.  They should be linked by
+                //     pointers only anyway, but the findUnits function is still testing names and equivalence ...
+                return t == u || u->name() == t->name();
```
Replace with `return (t == u) || (u->name() == t->name()); `.

kerimoyle replied:
f15e055d

kerimoyle replied:
Removed this function in 8837c78c

agarny commented on tests/model/model.cpp:
```diff
@@ -624,3 +624,59 @@ TEST(Model, removeComponentInsensitiveToOrder)
+TEST(Model, cleanModel)
+{
+    // Make a model with empty components, unrequired units, empty import sources.
+    auto e = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
```
Replace `auto e` with `const std::string e`. Same elsewhere in this PR.

kerimoyle replied:
f15e055d

agarny reviewed (commented):
Some initial comments until we come to some agreement regarding this PR. :)

nickerso commented on src/api/libcellml/model.h:
```diff
@@ -414,17 +414,25 @@ class LIBCELLML_EXPORT Model: public ComponentEntity
+     * @brief Remove any empty import sources, empty components, and unused units from the 
+     *        model.
+     * 
+     *  Remove any empty import sources, empty components, and unused units from the model.
```
Need to define what empty means in this context.

kerimoyle replied:
f15e055d

nickerso commented on src/model.cpp:
```diff
@@ -547,4 +547,72 @@ bool Model::fixVariableInterfaces()
 
+static const size_t MAX_SIZE_T = std::numeric_limits<size_t>::max();
+
+bool isComponentEmpty(const ComponentPtr &component)
```
I'd rename this something like `findAndRemoveEmptyComponents`

kerimoyle replied:
f15e055d

nickerso commented on src/model.cpp:
```diff
@@ -547,4 +547,72 @@ bool Model::fixVariableInterfaces()
+           && !component->isImport();
+}
+
+void checkUnits(const ComponentPtr &component, std::vector<UnitsPtr> &unused)
```
This method would also need to check if the units are used in the math (including math in the resets)

kerimoyle replied:
You're right, bother.  In order to do that, I need to parse the MathML.  If the MathML isn't valid then it can't be parsed, so errors need to be logged.  There is no logger because this is done as part of the `Model` class.  I can move this into the `Printer` class, since that's where people will be using it anyway, and log errors there.  Thoughts?

nickerso replied:
I can almost see this as an optional switch on `printModel` to clean and then print. But might also be easier to simply not do anything with units. Just clean "empty" things for now, cleaning "unused" things can be a future addition if it would ever be useful.

kerimoyle replied:
In the first round of discussions we talked about that - and concluded somewhere that this should be something separate from the printer in the `Model` class.  I can't remember why!  If I don't check units then it can stay where it is in the `Model`, where it's a little more flexible as you're not constrained to printing (even if that's what most people will do).

kerimoyle replied:
8837c78c

kerimoyle committed:
Merge branch 'develop' into i775-model-clean-function

kerimoyle committed:
fixes for review

kerimoyle committed:
removed check for unused units

kerimoyle committed:
Merge branch 'develop' into i775-model-clean-function

nickerso committed:
updating to current main state

nickerso committed:
Merge branch 'main' of https://github.com/cellml/libcellml into model-clean

nickerso committed:
update model clean.

Updating the Model::clean() method to only clean empty components and units. Update tests accordingly.

nickerso committed:
Merge branch 'main' of https://github.com/cellml/libcellml into model-clean

hsorby committed:
Update model.i

hsorby committed:
Remove unnecessary cast.

hsorby committed:
Remove debug print statement in test.

hsorby committed:
Remove debug statement from cleanCreatedModel.

nickerso reviewed (dismissed):
@hsorby and @agarny - I've updated this one to make it clear what I'm thinking cleaning means.

nickerso committed:
trying to improve coverage for model cleaning

nickerso committed:
Merge branch 'i775-model-clean-function' of https://github.com/kerimoyle/libcellml into model-clean

nickerso committed:
Extend cleaning test to ensure coverage of cleaning encapsulated empty components

agarny commented on src/api/libcellml/model.h:
```diff
@@ -413,17 +413,30 @@ class LIBCELLML_EXPORT Model: public ComponentEntity
+     * 
+     *  Remove any empty components and units fom this model.
+     *  In this context, "empty" means:
+     *   - components with no name, id, resets, variables, maths, or non-empty child components; and
```
`non-empty child components`? Shouldn't it simply be `child components`?

nickerso replied:
I think this is right - its listing the things that an empty component does not have.

agarny commented on src/api/libcellml/model.h:
```diff
@@ -413,17 +413,30 @@ class LIBCELLML_EXPORT Model: public ComponentEntity
+     *  Remove any empty components and units fom this model.
+     *  In this context, "empty" means:
+     *   - components with no name, id, resets, variables, maths, or non-empty child components; and
+     *   - units which have no name, id, or child unit's.
```
`child units` rather than `child unit's`?

nickerso replied:
d1008b6

agarny commented on src/api/libcellml/model.h:
```diff
@@ -413,17 +413,30 @@ class LIBCELLML_EXPORT Model: public ComponentEntity
+     *   - units which have no name, id, or child unit's.
+     *  This is applied recursively. In the situation where a component is only called
+     *  non-empty because it has a child component, but that child is empty, both will be
+     *  removed.
```
Might want to reword this a bit to read `In the situation where a component has an empty child component, both will be removed. `.

nickerso replied:
25578c0

agarny commented on src/model.cpp:
```diff
@@ -527,6 +527,44 @@ bool Model::fixVariableInterfaces()
     return allOk;
 }
 
+static const size_t MAX_SIZE_T = std::numeric_limits<size_t>::max();
```
Might want to move this to `utilities.h` and also remove the one that is defined in `analyser.cpp`?

nickerso replied:
fcb43fc

nickerso reviewed (dismissed):
Now with full coverage.

agarny commented on src/model.cpp:
```diff
@@ -527,6 +527,44 @@ bool Model::fixVariableInterfaces()
+        }
+    }
+
+    // Remove empty units
```
Missing final point.

nickerso replied:
615f237

agarny commented on tests/bindings/python/test_model.py:
```diff
@@ -379,6 +379,41 @@ def test_imports(self):
+        self.assertEqual(4, model.componentCount())
+        self.assertEqual(3, model.unitsCount())
+
+        # Call the Model.clean() function to remove empty items.
```
Might want to mention that the empty items are `c4` and `u3`.

nickerso replied:
615f237

agarny commented on tests/model/model.cpp:
```diff
@@ -628,6 +628,106 @@ TEST(Model, removeComponentInsensitiveToOrder)
+    model->addComponent(c3);
+    model->addComponent(c4);
+
+    // test unnamed components in an encapsulation hierarchy
```
`// Test unnamed components in an encapsulation hierarchy.`

nickerso replied:
Tests changed, so that comment is no longer relevant.

agarny commented on tests/resources/dirty_model.cellml:
```diff
@@ -0,0 +1,13 @@
+    <component name="nonEmptyComponent">
+        <variable name="x" units="requiredUnits"/>
+    </component>
+  <component />
```
Formatting (to make it easier to read/review/maintain/etc.):
```xml
  <component name="nonEmptyComponent">
    <variable name="x" units="requiredUnits"/>
  </component>
  <component/>
```

nickerso replied:
615f237

agarny reviewed (dismissed):
A good start. A few minor comments and, more importantly, some missing tests, I think, when it comes to:
 - a unit definition that has child units, e.g.
 ```xml
<units name="child_unit"/>
<units name="empty_child_unit">
    <unit/>
</units>
<units name="my_unit">
    <unit units="child_unit"/>
    <unit units="empty_child_unit"/>
    <unit/>
</units>
 ```
 which, once cleaned, should give us:
 ```xml
<units name="child_unit"/>
<units name="my_unit">
    <unit units="child_unit"/>
</units>
 ```
 - unit not cleared if it is imported;
 - component not cleared if it has resets;
 - component not cleared if it has some math; and
 - component not cleared if it is imported.

nickerso committed:
refining tests for model cleaning to cover all cases

nickerso committed:
fixing typo picked up in review.

nickerso committed:
Clarify description of hierarchical component cleaning

nickerso committed:
Move definition of MAX_SIZE_T to common location in utilities.h

nickerso committed:
Cleaning up comments and formatting after review

hsorby commented on src/model.cpp:
```diff
@@ -527,6 +527,42 @@ bool Model::fixVariableInterfaces()
     return allOk;
 }
 
+bool findAndRemoveEmptyComponent(const ComponentPtr &component)
```
I really think this method should be called `traverseHierarchyAndRemoveIfEmpty`.

nickerso replied:
c1c7470 and 3514c3e

hsorby reviewed (changes requested):
The tests don't build up very well (or at all) but apart from the minor change requested I can accept this.

nickerso committed:
rename component cleaning method

nickerso committed:
keep clang format happy

nickerso committed:
and update the use of the component cleaning method

nickerso committed:
change to the more common version for spelling descendents

nickerso commented:
> "descendants" not "descendents". Also, may want to replace "cleaned" with "removed"?

descendents updated in 54cacbc

nickerso committed:
tidy up clean method description based on review feedback

nickerso commented:
> "descendants" not "descendents". Also, may want to replace "cleaned" with "removed"?

and "cleaned" part in db2331e
